### PR TITLE
Build to class refactor

### PR DESCRIPTION
**Pull request by johnjcsmith**
Refactor BK primitives:
- Standardise on non-plural function names for adding attributes (`step()` rather than `steps()` or `addStep()`)

### CODE REVIEW

**Review by jameslnewell — APPROVED** [review]
Not a big issue but do `command`, `parralelism` and `env` get shared with any other steps? They might not have needed a "helper", could maybe just inline them in the `CommandStep`.

File: src/builders/CommandStep.ts [hunk 1]
```diff
@@ -65,19 +72,70 @@ export class CommandStep implements StepBuilder, KeyBuilder, LabelBuilder, Condi
...
+    this.#soft_fail = fail
+    return this
+  }
+
+  timeout(minutes: number): this {
+    this.#timeout_in_minutes = minutes
+    return this
+  }
+
+  artifactPath(path: string): this {
```
Comment (jameslnewell) [inline src/builders/CommandStep.ts:110]: nit/ I reckon we should leave this out and in the future add a plugin builder instead to canonicalise around https://github.com/buildkite-plugins/artifacts-buildkite-plugin

File: src/builders/CommandStep.ts [hunk 2]
```diff
@@ -65,19 +72,70 @@ export class CommandStep implements StepBuilder, KeyBuilder, LabelBuilder, Condi
...
+  
+  env(env: Record<string, string | number>): this {
+    this.#env = env
+    return this
+  }
 
-  build(): CommandStepSchema {
-    if (!this.#command) {
-      throw new Error('CommandStep requires a command.')
+  addEnv(key: string, value: string | number): this {
```
Comment (jameslnewell) [inline src/builders/CommandStep.ts:86]: thoughts around changing this to singular `env(key, value)` and removing the existing `env(object)` for consistency with the singular pattern?

File: src/builders/CommandStep.ts [hunk 3]
```diff
@@ -65,19 +72,70 @@ export class CommandStep implements StepBuilder, KeyBuilder, LabelBuilder, Condi
...
-      command: this.#command,
+      ...{ [commandKey]: this.#command },
       ...this.#keyHelper.build(),
       ...this.#labelHelper.build(),
       ...this.#conditionHelper.build(),
       ...this.#branchesHelper.build(),
       ...this.#dependenciesHelper.build(),
       ...this.#skipHelper.build(),
+      ...(this.#parallelism ? { parallelism: this.#parallelism } : {}),
+      ...(this.#concurrency ? {concurrency: this.#concurrency} : {}),
```
Comment (jameslnewell) [inline src/builders/CommandStep.ts:129]: ☝️ for these numeric values, is `0` a valid value we need to handle?

File: src/builders/CommandStep.ts [hunk 4]
```diff
@@ -65,19 +72,70 @@ export class CommandStep implements StepBuilder, KeyBuilder, LabelBuilder, Condi
...
+      ...(this.#concurrency_group ? {concurrency_group: this.#concurrency_group} : {}),
+      ...(this.#env ? {env: this.#env} : {}),
+      ...(this.#soft_fail ? {soft_fail: this.#soft_fail} : {}),
+      ...(this.#timeout_in_minutes ? {timeout_in_minutes: this.#timeout_in_minutes} : {}),
+      ...(this.#artifact_paths ? {artifact_paths: this.#artifact_paths} : {}),
+    }
+
+    if (!this.#command || (Array.isArray(this.#command) && this.#command.length === 0)) {
+      throw new Error('CommandStep requires a command.')
```
Comment (jameslnewell) [inline src/builders/CommandStep.ts:138]: Definitely what the docs suggest but does it actually? Isn't there times we've just used the docker plugin command?

**Review by jameslnewell — APPROVED** [review]

File: src/builders/CommandStep.ts (hunk 1 above)
Comment (johnjcsmith) [reply src/builders/CommandStep.ts]: ✅

File: src/builders/CommandStep.ts (hunk 2 above)
Comment (johnjcsmith) [reply src/builders/CommandStep.ts]: ✅

File: src/builders/CommandStep.ts (hunk 3 above)
Comment (johnjcsmith) [reply src/builders/CommandStep.ts:115]: 🤦

File: src/builders/CommandStep.ts (hunk 4 above)
Comment (johnjcsmith) [reply src/builders/CommandStep.ts]: Removing this for now

File: src/builders/CommandStep.ts (hunk 3 above)
Comment (johnjcsmith) [reply src/builders/CommandStep.ts:115]: Actually, both provide behaviour that I can't imagine would ever be used intentionally:
- `parallelism: 0` just removes the step
- `concurrency: 0` just causes the step to hang

Comment (johnjcsmith) [reply src/builders/CommandStep.ts]: Added errors for both cases

Comment (jameslnewell) [reply src/builders/CommandStep.ts]: Twas just checking cause `0 == false` and would result in `{}` when the values are `0`

**Pull request merged**